src/sync/partial_sync.cpp

```diff
@@ -0,0 +1,124 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "partial_sync.hpp"
+
+#include "impl/notification_wrapper.hpp"
+#include "impl/object_accessor_impl.hpp"
+#include "object_schema.hpp"
+#include "results.hpp"
+#include "shared_realm.hpp"
+#include "thread_safe_reference.hpp"
+
+#include <realm/util/scope_exit.hpp>
+
+namespace realm {
+namespace partial_sync {
+
+namespace {
+
+constexpr const char* result_sets_type_name = "ResultSets";
+
+std::string matches_property_name_for_object(const std::string& name)
+{
+    return name + "_matches";
+}
+
+Schema add_result_sets_to_schema(const Schema& source_schema, Property matches_property)
+{
+    std::vector<ObjectSchema> schema;
+    schema.reserve(source_schema.size());
+    std::copy(source_schema.begin(), source_schema.end(), std::back_inserter(schema));
+
+    ObjectSchema result_sets_schema(result_sets_type_name, {
+        {"matches_property", PropertyType::String},
+        {"query", PropertyType::String},
+        {"status", PropertyType::Int},
+        {"error_message", PropertyType::String},
+        std::move(matches_property),
+    });
+    schema.push_back(std::move(result_sets_schema));
+    return schema;
+}
+
+} // unnamed namespace
+
+void register_query(std::shared_ptr<Realm> realm, const std::string &object_class, const std::string &query,
+                    std::function<void (Results, std::exception_ptr)> callback)
+{
+    auto matches_property = matches_property_name_for_object(object_class);
+
+    auto schema = add_result_sets_to_schema(realm->schema(),
+                                            {matches_property, PropertyType::Object|PropertyType::Array, object_class});
+
+    Object raw_object;
+    {
+        realm->begin_transaction();
+        auto cleanup = util::make_scope_exit([&]() noexcept {
+            if (realm->is_in_transaction())
+                realm->cancel_transaction();
+        });
+
+        // FIXME: Is it ok for us to modify the schema of the Realm like this?
```

Comment:
@tgoyne, can you please weigh in on the question asked in this FIXME?

Reply:
Seems likely to be a problem. IMO `__ResultsSet` should not be exposed in the Realm's schema (even for dynamic realms), and the schema reading code should be modified to skip it.

A drycoded idea:
```
    auto& group = realm.read_group();
    ObjectSchema result_sets_schema(result_sets_type_name, {
        {"matches_property", PropertyType::String},
        {"query", PropertyType::String},
        {"status", PropertyType::Int},
        {"error_message", PropertyType::String},
        {matches_property, PropertyType::Object|PropertyType::Array, object_class}
    });
    auto required_changes = Schema{{ObjectSchema{group, result_sets_type_name}}}.compare(Schema{{results_set_schema}})
    if (!required_changes.empty())
        apply_additive_changes(group, changes_required);
```

Reply:
Thanks for the suggestion. I'll give it a shot.

> IMO __ResultsSet should not be exposed in the Realm's schema (even for dynamic realms), and the schema reading code should be modified to skip it.

I agree with this, but some people have expressed a desire for it to be visible in the browser.

Reply:
Yes, specifically @bigfish24 has.

Reply:
I think it is valuable right now during the tech preview but maybe at GA we are more confident that we can hide it. I think it would be better to debug with a customer by having it available for the time being.